independent-projects/bootstrap/runner/src/main/java/io/quarkus/bootstrap/runner/JarResource.java

```diff
@@ -152,50 +153,36 @@ public ProtectionDomain getProtectionDomain() {
     }
 
     private JarFile readLockAcquireAndGetJarReference() {
-        while (true) {
-            readLock.lock();
-            final JarFile zipFileLocal = this.zipFile;
-            if (zipFileLocal != null) {
-                //Expected fast path: returns a reference to the open JarFile while owning the readLock
-                return zipFileLocal;
-            } else {
-                //This Lock implementation doesn't allow upgrading a readLock to a writeLock, so release it
-                //as we're going to need the WriteLock.
-                readLock.unlock();
-                //trigger the JarFile being (re)opened.
-                ensureJarFileIsOpen();
-                //Now since we no longer own any lock, we need to try again to obtain the readLock
-                //and check for the reference still being valid.
-                //This exposes us to a race with closing the just-opened JarFile;
-                //however this should be extremely rare, so we can trust we won't loop much;
-                //A counter doesn't seem necessary, as in fact we know that methods close()
-                //and resetInternalCaches() are invoked each at most once, which limits the amount
-                //of loops here in practice.
+        while (zipFileLoadingState.get() != ZipFileLoadingState.LOADED) {
+            if (zipFileLoadingState.compareAndSet(ZipFileLoadingState.NOT_LOADED, ZipFileLoadingState.LOADING)) {
+                return loadZipFile();
             }
+            Thread.yield();
         }
+        zipFileReadersCounter.incrementAndGet();
```

Comment:
Here a close and happen (and a subsequent open too!!!) which means that we can have the zip file in 2 wrong states:
- already closed
- a different one(!)

I suggest to move the ref cnt into a atomic ref < ref cnt obj > which keep the 2 states dependent and tied together